include/ccf/pal/attestation.h

```diff
@@ -2,22 +2,108 @@
+#  include "ccf/crypto/pem.h"
+#  include "ccf/crypto/verifier.h"
+#  include "ccf/pal/attestation_sev_snp.h"
+#  include "clients/rpc_tls_client.h"
+#  include "crypto/ecdsa.h"
+
+#  include <fcntl.h>
+#  include <sys/ioctl.h>
+#  include <unistd.h>
 #else
+#  include "ccf/pal/attestation_sgx.h"
+
 #  include <openenclave/attestation/attester.h>
 #endif
 
 namespace ccf::pal
 {
 #if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
 
-  static QuoteInfo generate_quote(attestation_report_data&&)
+  static QuoteInfo generate_quote(attestation_report_data& report_data)
   {
     QuoteInfo node_quote_info = {};
-    node_quote_info.format = QuoteFormat::insecure_virtual;
+    auto is_sev_snp = access(snp::DEVICE, F_OK) == 0;
+
+    // If there is no SEV-SNP device, assume we are using insecure virtual
+    // quotes
+    if (!is_sev_snp)
+    {
+      node_quote_info.format = QuoteFormat::insecure_virtual;
+    }
+
+    else
+    {
+      node_quote_info.format = QuoteFormat::amd_sev_snp_v1;
+      int fd = open(snp::DEVICE, O_RDWR | O_CLOEXEC);
+      if (fd < 0)
+      {
+        throw std::logic_error(
+          fmt::format("Failed to open \"{}\"", snp::DEVICE));
+      }
+
+      snp::AttestationReq req = {};
+      snp::AttestationResp resp = {};
+
+      // Arbitrary report data
+      memcpy(req.report_data, report_data.data(), attestation_report_data_size);
+
+      // Documented at
+      // https://www.kernel.org/doc/html/latest/virt/coco/sev-guest.html
+      snp::GuestRequest payload = {
+        .req_msg_type = snp::MSG_REPORT_REQ,
+        .rsp_msg_type = snp::MSG_REPORT_RSP,
+        .msg_version = 1,
+        .request_len = sizeof(req),
+        .request_uaddr = reinterpret_cast<uint64_t>(&req),
+        .response_len = sizeof(resp),
+        .response_uaddr = reinterpret_cast<uint64_t>(&resp),
+        .error = 0};
+
+      int rc = ioctl(fd, SEV_SNP_GUEST_MSG_REPORT, &payload);
+      if (rc < 0)
+      {
+        CCF_APP_FAIL("IOCTL call failed: {}", strerror(errno));
+        CCF_APP_FAIL("Payload error: {}", payload.error);
+        throw std::logic_error(
+          "Failed to issue ioctl SEV_SNP_GUEST_MSG_REPORT");
+      }
+
+      auto quote = &resp.report;
+      auto quote_bytes = reinterpret_cast<uint8_t*>(&resp.report);
+      node_quote_info.quote.assign(quote_bytes, quote_bytes + resp.report_size);
+
+      client::RpcTlsClient client{
+        "americas.test.acccache.azure.net",
+        "443",
+        nullptr,
+        std::make_shared<tls::Cert>(
+          nullptr, std::nullopt, std::nullopt, std::nullopt, false)};
+
+      auto params = nlohmann::json::object();
+      params["api-version"] = "2020-10-15-preview";
+
+      auto response = client.get(
+        fmt::format(
+          "/SevSnpVM/certificates/{}/{}",
+          fmt::format("{:02x}", fmt::join(quote->chip_id, "")),
+          fmt::format("{:0x}", *(uint64_t*)(&quote->reported_tcb))),
+        params);
```

Comment:
This client code was previously only used in untrusted clients, so this is the first thing that brings it into the TCB. The underlying crypto bits are shared with the already trusted implementation, so probably fine, but we need to be aware of this change. This is also a blocking call making a network request, with no timeout. We may need to add a timeout, or convert this entire `generate_quote` call to something async to handle implementations where it is inherently async. Fine for this first-pass PR, but I think we should follow up with another examination/hardening pass.

Reply:
Sounds like we can do it alongside making this whole thing configurable?

Reply:
Side note: that doesn't fundamentally need to be trusted, it could be done on the host side, before starting the enclave. In Parma, there is no such distinction, but there will be in Monza, and so could be on that side there.